sql: add inverted index validation after backfill

thoszhang opened the pull request:
This change adds a validation step after backfilling an inverted index, before
the index is made public. The validation is performed by counting the number of
index entries (using a KV scan) and comparing the count to the expected number
of index entries, which is computed from a newly added builtin function.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/35056)
<!-- Reviewable:end -->

lucy-zhang commented:
I ran the roachtest in #35058 on a table with 5 million rows and 30 million index entries, and the index was built in 16m34s, including validation which took 3m28s (21% of the total time). The performance of reads and writes was really degraded, though---here are stats for updates/selects each on a single key:
```
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
 1200.0s        0         175347          146.1      2.1      1.4      5.0     10.5   2013.3  read
 1200.0s        0         174557          145.5     52.9     17.8     41.9   1342.2  22548.6  write

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
 1200.0s        0         349904          291.6     27.4      7.6     31.5    104.9  22548.6
```
When I ran the roachtest for ~30 million rows (which is an hour's worth of generated data), the test hung, and when I killed the test, it also wiped the cluster (?). I'll have to try to reproduce again.

awoods187 commented:
you might have to pass the --wipe=false flag

lucy-zhang commented:
I had `--debug` set, which disables wiping the cluster after a test failure (I think `--wipe=false` is for disabling wiping the cluster before the test starts). It might have been something else I did.

dt reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35056)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @lucy-zhang)

___
*[pkg/sql/backfill.go, line 687 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35056#-LZC0p6S5NhKJ5RoChTC:-LZC0p6S5NhKJ5RoChTD:b-69xixv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d553da0afae062e3d21711c516751bee34f06a50/pkg/sql/backfill.go#L687)):*
> ```Go
> 
> 		grp.GoCtx(func(ctx context.Context) error {
> 			// Inverted indexes can't be interleaved, so a KV scan can be used to get the index length.
> ```

I might say this isn't something we want to count on long term -- we're breaking some abstractions here about SQL encoding. Additionally, I know there's some overhead with SQL flows, but _in theory_ pulling everything to one node to count it _should_ be slower than pushing counters out to every node.

I'd probably say this deserves a TODO to switch to proper SQL decoding and execution via dist-sql. 


<!-- Sent from Reviewable.io -->

dt reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35056)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @lucy-zhang)

___
*[pkg/sql/sem/builtins/builtins.go, line 3007 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35056#-LZC4hK1CzyKrjJCRUa_:-LZC4hK24xqz9SLuhIbx:b-5z65oq) ([raw file](https://github.com/cockroachdb/cockroach/blob/d553da0afae062e3d21711c516751bee34f06a50/pkg/sql/sem/builtins/builtins.go#L3007)):*
> ```Go
> 			ReturnType: tree.FixedReturnType(types.Int),
> 			Fn: func(_ *tree.EvalContext, args tree.Datums) (tree.Datum, error) {
> 				keys, err := json.EncodeInvertedIndexKeys(make([]byte, 0), tree.MustBeDJSON(args[0]).JSON)
> ```

LGTM for now. I've love to find a way to not alloc every path.
minor nit: could move the meat of this function into a func in `json` so iterating on it later is more contained. 


<!-- Sent from Reviewable.io -->

lucy-zhang committed:
sql: add inverted index validation after backfill

This change adds a validation step after backfilling an inverted index, before
the index is made public. The validation is performed by counting the number of
index entries (using a KV scan) and comparing the count to the expected number
of index entries, which is computed from a newly added builtin function.

Release note: None

thoszhang reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35056)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt)

___
*[pkg/sql/backfill.go, line 687 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35056#-LZC0p6S5NhKJ5RoChTC:-LZGdkbQ1UzRwHuz31Zf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d553da0afae062e3d21711c516751bee34f06a50/pkg/sql/backfill.go#L687)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I might say this isn't something we want to count on long term -- we're breaking some abstractions here about SQL encoding. Additionally, I know there's some overhead with SQL flows, but _in theory_ pulling everything to one node to count it _should_ be slower than pushing counters out to every node.

I'd probably say this deserves a TODO to switch to proper SQL decoding and execution via dist-sql. 
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/builtins.go, line 3007 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35056#-LZC4hK1CzyKrjJCRUa_:-LZGdl5w6458N5dIY2rP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d553da0afae062e3d21711c516751bee34f06a50/pkg/sql/sem/builtins/builtins.go#L3007)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

LGTM for now. I've love to find a way to not alloc every path.
minor nit: could move the meat of this function into a func in `json` so iterating on it later is more contained. 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

lucy-zhang commented:
bors r+
